sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -998,6 +998,22 @@ class Analyzer(
   }
 
   /**
+   * Replace unresolved expressions in grouping keys with resolved ones in SELECT clauses.
+   */
+  object ResolveAggAliasInGroupBy extends Rule[LogicalPlan] {
+
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperators {
+      case agg @ Aggregate(groups, aggs, child)
+          if conf.groupByAliases && child.resolved && aggs.forall(_.resolved) &&
```

Comment:
The latest fix is not enough for your suggestion?;
This fix checks if `UnresolvedAttribute` exists there, then filter out by `child.output`.
https://github.com/apache/spark/pull/17191/files#diff-57b3d87be744b7d79a9beacf8e5e5eb2R1014
```
 +        // This is a strict check though, we put this to apply the rule only in alias expressions
 +        def checkIfChildOutputHasNo(attrName: String): Boolean =
 +          !child.output.exists(a => resolver(a.name, attrName))
 +        agg.copy(groupingExpressions = groups.map {
 +          case u: UnresolvedAttribute if checkIfChildOutputHasNo(u.name) =>
 +            aggs.find(ne => resolver(ne.name, u.name)).getOrElse(u)
 +          case e => e
 +        })
```